embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/resurrection/PayloadResurrectionServiceImpl.kt

```diff
@@ -0,0 +1,149 @@
+package io.embrace.android.embracesdk.internal.resurrection
+
+import io.embrace.android.embracesdk.internal.arch.schema.EmbType
+import io.embrace.android.embracesdk.internal.clock.nanosToMillis
+import io.embrace.android.embracesdk.internal.delivery.StoredTelemetryMetadata
+import io.embrace.android.embracesdk.internal.delivery.SupportedEnvelopeType
+import io.embrace.android.embracesdk.internal.delivery.intake.IntakeService
+import io.embrace.android.embracesdk.internal.delivery.storage.PayloadStorageService
+import io.embrace.android.embracesdk.internal.logging.EmbLogger
+import io.embrace.android.embracesdk.internal.logging.InternalErrorType
+import io.embrace.android.embracesdk.internal.ndk.NativeCrashService
+import io.embrace.android.embracesdk.internal.opentelemetry.embCrashId
+import io.embrace.android.embracesdk.internal.opentelemetry.embHeartbeatTimeUnixNano
+import io.embrace.android.embracesdk.internal.payload.Attribute
+import io.embrace.android.embracesdk.internal.payload.Envelope
+import io.embrace.android.embracesdk.internal.payload.NativeCrashData
+import io.embrace.android.embracesdk.internal.payload.SessionPayload
+import io.embrace.android.embracesdk.internal.payload.Span
+import io.embrace.android.embracesdk.internal.payload.getSessionSpan
+import io.embrace.android.embracesdk.internal.payload.toFailedSpan
+import io.embrace.android.embracesdk.internal.serialization.PlatformSerializer
+import io.embrace.android.embracesdk.internal.spans.findAttributeValue
+import io.embrace.android.embracesdk.internal.spans.hasFixedAttribute
+import io.embrace.android.embracesdk.internal.utils.Provider
+import io.opentelemetry.semconv.incubating.SessionIncubatingAttributes
+import java.util.zip.GZIPInputStream
+import kotlin.math.max
+
+internal class PayloadResurrectionServiceImpl(
+    private val intakeService: IntakeService,
+    private val payloadStorageService: PayloadStorageService,
+    private val nativeCrashServiceProvider: Provider<NativeCrashService?>,
+    private val logger: EmbLogger,
+    private val serializer: PlatformSerializer,
+) : PayloadResurrectionService {
+
+    override fun resurrectOldPayloads() {
+        payloadStorageService
+            .getUndeliveredPayloads()
+            .forEach { deadSessionMetadata ->
+                deadSessionMetadata.sendResurrectedPayload()
+            }
+    }
+
+    /**
+     * Load and modify the given incomplete payload envelope and send the result to the [IntakeService] for delivery.
+     * Resurrected payloads sent to the [IntakeService] will be deleted.
+     */
+    private fun StoredTelemetryMetadata.sendResurrectedPayload() {
+        val result = runCatching {
+            val deadPayload = when (envelopeType) {
+                SupportedEnvelopeType.SESSION -> {
+                    val deadSession = serializer.fromJson<Envelope<SessionPayload>>(
+                        inputStream = GZIPInputStream(payloadStorageService.loadPayloadAsStream(this)),
```

Comment:
Agreed. Ran into this when I was doing the fakes. Let me think about how I can add on to this (after the merge)

Reply:
Per our discussion, I'm going to document that the bytes being put into the PayloadStorageService will be gzipped, and the bytes coming out will therefore always be gzipped. This is so that when we use it in the payload of the request, we don't have to re-zip it. 

If we ever need to support a case where we get the exact bytes back (i.e. do decompression first so the bytestream is not zipped), we can either add a a parameter to the load method, or have another one.